src/libostree/ostree-sysroot.c

```diff
@@ -1738,9 +1743,15 @@ ostree_sysroot_simple_write_deployment (OstreeSysroot      *sysroot,
   if (!booted_deployment && !merge_deployment && (retain_pending || retain_rollback))
     retain = TRUE;
 
+  /* tracks current versioned deployment */
+  OstreeRepo *repo = ostree_sysroot_repo (sysroot);
+  const gchar *last_version =
+    _ostree_deployment_get_version (new_deployment, repo, error);
```

Comment:
I think here, we should also check if the `booted_deployment` or `merge_deployment` have different versions from `new_version`. We never prune those, so it would fulfill the criteria. So then, we'd set `retained_previous_version = TRUE` here.

If they *aren't* different (or there is no booted/merge), then the logic below is good.